### PR TITLE
Fix spray-json unmarshaller utf8 encoding bug

### LINKED ISSUE #18389: spray-json support unmarshaller charset problem
Hi,

I have experienced a strange problem and it seems to be a bug. Below is the request json that I send to a service:

```json
{"title":"ÇatÇat Kitapları","parentName":"Çocuk Kitapları"}
```

This json is decoded as:

```json
{"title":"Ã�atÇat Kitapları","parentName":"Ã�ocuk Kitapları"}
```

But more interestingly when I put a character before `Ç` as below:

```json
{"title":"aÇatÇat Kitapları","parentName":"aÇocuk Kitapları"}
```

It is decoded perfectly as below:

```json
{"title":"aÇatÇat Kitapları","parentName":"aÇocuk Kitapları"}
```

As I saw some very strange things are happening here I tried to debug the problem. Below is the code snippet that seems to be responsible for JsonParser creation in SprayJsonSupport.scala

```scala
Unmarshaller.byteStringUnmarshaller.forContentTypes(`application/json`).mapWithCharset { (data, charset) ⇒
  val input =
    if (charset == HttpCharsets.`UTF-8`) ParserInput(data.toArray)
    else ParserInput(data.decodeString(charset.nioCharset.name)) // FIXME: identify charset by instance, not by name!
  JsonParser(input)
}
```

So I tried changing the related line to this:

```scala
 if (charset == HttpCharsets.`UTF-8`) ParserInput(data.utf8String)
```

And everything worked perfectly. So am I doing smth wrong or is there a problem in spray-json or maybe in ByteString?

### PR DESCRIPTION
**Pull request by fehmicansaglam**
Fixes #18389

### CODE REVIEW

File: akka-http-tests/src/test/scala/akka/http/scaladsl/marshallers/JsonSupportSpec.scala [hunk 1]
```diff
@@ -19,11 +19,11 @@ object Employee {
   val simple = Employee("Frank", "Smith", 42, 12345, false)
   val json = """{"fname":"Frank","name":"Smith","age":42,"id":12345,"boardMember":false}"""
 
-  val utf8 = Employee("Fränk", "Smi√", 42, 12345, false)
+  val utf8 = Employee("Fränk", "Çmi√", 42, 12345, false)
   val utf8json =
     """{
       |  "fname": "Fränk",
-      |  "name": "Smi√",
```
Comment (ktoso) [inline akka-http-tests/src/test/scala/akka/http/scaladsl/marshallers/JsonSupportSpec.scala]: thanks for amending the test to cover this case!

**Comment by ktoso** [comment]
Thanks for discovering the problem and fix @fehmicansaglam!
Can you amend the commit to have its message abide our guidelines ( https://github.com/akka/akka/blob/master/CONTRIBUTING.md#creating-commits-and-writing-commit-messages ) ? Here it would be `=htp #18389 Fix spray-json unmarshaller utf8 encoding bug`

**Comment by ktoso** [comment]
OK TO TEST

**Comment by fehmicansaglam** [comment]
@ktoso done.

**Comment by ktoso** [comment]
LGTM, thanks! Will merge after PR validation

**Pull request merged**